src/libpmemfile/preload.c

```diff
@@ -0,0 +1,2004 @@
+ * whenever these file descriptors are encountered during interposing, -EBADF
+ * must be returned. The contents of this array does not change after startup.
+ */
+static bool mount_point_fds[PMEMFILE_MAX_FD + 1];
+
+static struct fd_association fd_table[PMEMFILE_MAX_FD + 1];
+
+static pthread_rwlock_t pmem_cwd_lock = PTHREAD_RWLOCK_INITIALIZER;
+static struct pool_description *volatile cwd_pool;
+
+static int exit_on_ENOTSUP;
+static long check_errno(long e)
+{
+	if (e == -ENOTSUP && exit_on_ENOTSUP) {
+		const char *str = "ENOTSUP";
+
+		syscall_no_intercept(SYS_write, 2, str, strlen(str));
+		exit_group_no_intercept(95);
+	}
+
+	return e;
+}
+
+static struct fd_desc
+cwd_desc()
+{
+	struct fd_desc result;
+
+	result.kernel_fd = AT_FDCWD;
+	result.pmem_fda.pool = cwd_pool;
+	result.pmem_fda.file = PMEMFILE_AT_CWD;
+
+	return result;
+}
+
+static struct fd_desc
+fetch_fd(long fd)
+{
+	struct fd_desc result;
+
+	result.kernel_fd = fd;
+
+	if ((int)fd == AT_FDCWD) {
+		result.pmem_fda.pool = cwd_pool;
+		result.pmem_fda.file = PMEMFILE_AT_CWD;
+	} else if (fd_pool_has_allocated(fd)) {
+		result.pmem_fda = fd_table[fd];
+	} else {
+		result.pmem_fda.pool = NULL;
+		result.pmem_fda.file = NULL;
+	}
+
+	return result;
+}
+
+static __thread bool reenter = false;
+/*
+ * This way the default can be overridden from the command line during
+ * a build, without altering the source. As a second option, it can be
+ * overridden using a environment variable at runtime.
+ */
+#ifndef PMEMFILE_DEFAULT_USE_SYMLINK_STRICT
+#define PMEMFILE_DEFAULT_USE_SYMLINK_STRICT false
+#endif
+
+static bool use_stricter_symlink_resolver = PMEMFILE_DEFAULT_USE_SYMLINK_STRICT;
+
+static void log_init(const char *path, const char *trunc);
+static void log_write(const char *fmt, ...) pf_printf_like(1, 2);
+
+static void establish_mount_points(const char *);
+static void setup_strict_symlink_flag(const char *);
+static void init_hooking(void);
+
+pf_constructor void
+pmemfile_preload_constructor(void)
+{
+	if (!libc_hook_in_process_allowed())
+		return;
+
+	log_init(getenv("PMEMFILE_PRELOAD_LOG"),
+			getenv("PMEMFILE_PRELOAD_LOG_TRUNC"));
+
+	const char *env_str = getenv("PMEMFILE_EXIT_ON_NOT_SUPPORTED");
+	exit_on_ENOTSUP = env_str ? env_str[0] == '1' : 0;
+
+	// establish_mount_points already needs to have the flag set up
+	setup_strict_symlink_flag(getenv("PMEMFILE_USE_SYMLINK_STRICT"));
+
+	establish_mount_points(getenv("PMEMFILE_POOLS"));
+
+	if (pool_count == 0)
+		return; // No pools mounted. TODO: prevent syscall interception
+
+	/*
+	 * Must be the last step, the callback can be called anytime
+	 * after the call to init_hooking()
+	 */
+	init_hooking();
+}
+
+static void
+setup_strict_symlink_flag(const char *e)
+{
+	// Only overwrite the default when explicitly requested by an env var.
+	if (e != NULL)
+		use_stricter_symlink_resolver = (e[0] != '0');
+}
+
+static void
+init_hooking(void)
+{
+	// todo: move this filtering to the intercepting library
```

Comment:
maybe instead of having three array lookups, you could define your own STATIC array of structs: 
```
{
 bool requires_wlock;
 bool requires_cwd_wlock;
 ...
}
```